src/compile/data/parse.ts

```diff
@@ -322,6 +322,7 @@ export function parseData(model: Model): DataComponent {
   }
 
   head = ParseNode.makeImplicitFromEncoding(head, model, ancestorParse) || head;
```

Comment:
I believe that we may create unnecessarily flattening for fields that are already flattened in `makeImplicitFromEncoding`. I see three solutions. First, we could extract the code that computes `implicit` for encodings and selections and then call `makeWithAncestors` once for a single implicit dict. Second, makeImplicitFromEncoding could return its `implicit`, which we pass to `makeImplicitFromSelection`, and there we ignore fields that are already flattened otherwise. Second, you could merge `makeImplicitFromEncoding` and `makeImplicitFromSelection` so that they share the same `implicit` dict. I guess this is like the first approach but encapsulates the logic. I'd prefer the first as it's most explicit.

Reply:
Thanks for the comment. Yes, I'd thought about this and tested it informally (e.g., using the spec below in which both `makeImplicitFromEncoding` and `makeImplicitFromSelection` are called). As far as I could tell, duplicate flattening does _not_ occur. But, I wasn't sure how to formally test this as a unit test (i.e., focused on just parsing/assembling the `ParseNode`). I'm happy to add a test case that parses the whole model and assembles the full dataflow and test what's in the `transform` array, but that doesn't feel like the right approach. Open to ideas/suggestions here. 

Re: your suggested solutions, option (1) won't work because the existing `fromEncoding` logic expects `TypedFieldDefs` -- selections only hold field and channel names as strings to simplify assembly. I'm happy to merge into a single `makeImplicitFromEncodingAndSelection` such that only a single `implicit` object is populated and passed to `makeWithAncestors` if we prefer that. I chose the current approach because I thought that better separated concerns. 

Let me know what you'd prefer.

Reply:
```json
{
  "$schema": "https://vega.github.io/schema/vega-lite/v3.json",
  "selection": {
    "grid": {
      "type": "multi", "fields": ["nested.a"]
    }
  },
  "data": {
    "values": [
      {"nested":{ "a": "1", "b": 28}},
      {"nested":{ "a": "2", "b": 55}},
      {"nested":{ "a": "3", "b": 43}},
      {"nested":{ "a": "4", "b": 91}},
      {"nested":{ "a": "5", "b": 81}},
      {"nested":{ "a": "6", "b": 53}},
      {"nested":{ "a": "7", "b": 19}},
      {"nested":{ "a": "8", "b": 87}},     
      {"nested":{ "a": "8", "b": 52}}
    ]
  },
  "mark": "point",
  "encoding": {
    "y": {
      "field": "nested.b",
      "type": "quantitative"
    },
    "x": {
      "field": "nested.b",
      "type": "quantitative"
    },
    "color": {
      "condition": {"selection": "grid", "value": "red"}
    }
  }
}
```

does not yield any duplicate flattening (though it should, were it an issue I think?).

Reply:
Which field would you expect to be flattened twice?

I can take care of avoiding duplicate flattening since I am more familiar with the dataflow stuff.

Reply:
Sorry, I think we did a simultaneous update. In the updated spec above, the selection is projected over `nested.b` which is also used in `encoding`. We would expect to see duplicates, but we don't. I think this was your original concern in https://github.com/vega/vega-lite/pull/5351#discussion_r321986174 but perhaps I'm misunderstanding?